File: src/main/scala/easy/No104_MaximumDepthOfBinaryTree.scala
package easy

import scala.util.Try

object No104_MaximumDepthOfBinaryTree {

  class TreeNode(var _value: Int) {
    var value: Int = _value
    var left: TreeNode = null
    var right: TreeNode = null
  }

  /*
  BFS广度优先
   */
  def maxDepth(root: TreeNode): Int = {
    if (root == null) {
       0
    } else {
      val queue=scala.collection.mutable.Queue[TreeNode]()
      var depth=0
      queue.enqueue(root)
      while(queue.nonEmpty){
        depth+=1
        for(i<- queue.indices){
          val son=queue.dequeue()
      if(son.left!=null)queue.enqueue(son.left)
          if(son.right!=null)queue.enqueue(son.right)
        }
      }
      depth
    }
  }

  /** *
    * 功能实现:DFS
    *
    * Author: Lzy
    * Date: 2018/12/10 9:12
    * Param: [root]
    * Return: int
    */
  def maxDepth2(root: TreeNode): Int = {
    if (root == null) 0
    else 1 + math.max(maxDepth(root.left), maxDepth(root.right))

    root match {
      case null => 0
      case _ => math.max(maxDepth(root.left), maxDepth(root.right)) + 1
    }
  }


  /** *
    * 功能实现:最小深度
    *
    * Author: Lzy
    * Date: 2018/12/10 9:12
    * Param: [root]
    * Return: int
    */
  def minDepth3(root: TreeNode): Int = {
    root match {
      case null =>0
      case _=>{
        val left=minDepth3(root.left)
        val right=minDepth3(root.right)
        //left+right+1不管哪个为0都可以直接使用，不然还需要单独判断
        if(left ==0 || right ==0) left+right+1 else math.min(left,right)+1
      }
    }
  }
}

File: src/main/java/binaryTree/No104_MaximumDepthOfBinaryTree.java
package binaryTree;

import java.util.ArrayDeque;
import java.util.Queue;

/**
 * 104. 二叉树的最大深度
 * 给定一个二叉树，找出其最大深度。
 * 二叉树的深度为根节点到最远叶子节点的最长路径上的节点数。
 * 说明: 叶子节点是指没有子节点的节点。
 * 示例：
 * 给定二叉树 [3,9,20,null,null,15,7]，
 * 3
 * / \
 * 9  20
 * /  \
 * 15   7
 * 返回它的最大深度 3 。
 * <p>
 * 111. 二叉树的最小深度
 * 给定一个二叉树，找出其最小深度。
 * 最小深度是从根节点到最近叶子节点的最短路径上的节点数量。
 * 说明：叶子节点是指没有子节点的节点。
 * 示例 1：
 * 输入：root = [3,9,20,null,null,15,7]
 * 输出：2
 * 示例 2：
 * 输入：root = [2,null,3,null,4,null,5,null,6]
 * 输出：5
 * 提示：
 * 树中节点数的范围在 [0, 105] 内
 * -1000 <= Node.val <= 1000
 *
 * @author: Patrick Star
 * @time: 2021/5/11 23:04
 */
public class No104_MaximumDepthOfBinaryTree {
    /**
     * 最大深度
     * 最快的
     *
     * @param root
     * @return
     */
    public int maxDepth(TreeNode root) {
        if (root == null) return 0;
        int leftDepth = maxDepth(root.left);
        int rightDepth = maxDepth(root.right);
        return Math.max(leftDepth, rightDepth) + 1;
    }

    /**
     * 最小深度
     *
     * @param root
     * @return
     */
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        if (root.left == null && root.right == null) return 1;
        int leftDepth = minDepth(root.left);
        int rightDepth = minDepth(root.right);
        if (root.left == null || root.right == null) return leftDepth + rightDepth + 1;
        return Math.min(leftDepth, rightDepth) + 1;
    }

    /**
     * 广度优先
     * 最快的
     *
     * @param root
     * @return
     */
    public int minDepth2(TreeNode root) {
        if (root == null) return 0;
        Queue<TreeNode> queue = new ArrayDeque<>(100);
        int level = 1;
        queue.add(root);
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode cur = queue.poll();
                if (cur.left == null && cur.right == null) return level;
                if (cur.left != null) queue.add(cur.left);
                if (cur.right != null) queue.add(cur.right);
            }
            level++;
        }
        return 0;

    }
}
